pkg/flow/internal/controller/loader.go

```diff
@@ -564,84 +568,106 @@ func (l *Loader) OriginalGraph() *dag.Graph {
-	level.Info(logger).Log("msg", "starting partial graph evaluation")
-	defer func() {
-		span.SetStatus(codes.Ok, "")
+	l.mut.RLock()
+	defer l.mut.RUnlock()
 
-		duration := time.Since(start)
-		level.Info(logger).Log("msg", "finished partial graph evaluation", "duration", duration)
-		l.cm.componentEvaluationTime.Observe(duration.Seconds())
-	}()
+	dependenciesToParentsMap := make(map[dag.Node]*ComponentNode)
+	for _, parent := range updatedNodes {
+		// Make sure we're in-sync with the current exports of parent.
+		l.cache.CacheExports(parent.ID(), parent.Exports())
+		// We collect all nodes directly incoming to parent.
+		_ = dag.WalkIncomingNodes(l.graph, parent, func(n dag.Node) error {
+			dependenciesToParentsMap[n] = parent
+			return nil
+		})
+	}
 
-	// Make sure we're in-sync with the current exports of c.
-	l.cache.CacheExports(c.ID(), c.Exports())
+	// Submit all dependencies for asynchronous evaluation.
+	// During evaluation, if a node's exports change, Flow will add it to updated nodes queue (controller.Queue) and
+	// the Flow controller will call EvaluateDependencies on it again. This results in a concurrent breadth-first
+	// traversal of the nodes that need to be evaluated.
+	for n, parent := range dependenciesToParentsMap {
+		dependantCtx, span := tracer.Start(spanCtx, "SubmitForEvaluation", trace.WithSpanKind(trace.SpanKindInternal))
+		span.SetAttributes(attribute.String("node_id", n.NodeID()))
+		span.SetAttributes(attribute.String("originator_id", parent.NodeID()))
 
-	_ = dag.WalkReverse(l.graph, c, func(n dag.Node, parent dag.Node) error {
-		if n == c {
-			// Skip over the starting component; the starting component passed to
-			// EvaluateDependencies had its exports changed and none of its input
-			// arguments will need re-evaluation.
-			return nil
+		// Submit the node for asynchronous evaluation.
+		err := l.workerPool.SubmitWithKey(n.NodeID(), l.concurrentEvalFn(n, dependantCtx, tracer, parent))
+		if err != nil {
+			// The error typically means that the workerPool queue is full. This could mean we have too many components
+			// and the agent cannot keep up with the evaluation. To degrade gracefully, we log the error and continue.
+			// Some evaluations may be missed.
+			level.Error(l.log).Log("msg", "failed to submit node for evaluation - the agent is likely "+
+				"overloaded and cannot keep up with evaluating components - some updates may be dropped",
+				"node_id", n.NodeID(), "err", err)
+			span.SetStatus(codes.Error, err.Error())
 		}
+		span.End()
+	}
+
+	// Report queue size metric.
+	l.cm.evaluationQueueSize.Set(float64(l.workerPool.QueueSize()))
+}
 
+// concurrentEvalFn returns a function that evaluates a node and updates the cache. This function can be submitted to
+// a worker pool for asynchronous evaluation.
+func (l *Loader) concurrentEvalFn(n dag.Node, spanCtx context.Context, tracer trace.Tracer, parent *ComponentNode) func() {
```

Comment:
style nit: It feels weird to me for this to return a function, instead of the caller wrapping it to make it async:

```go
// Submit the node for asynchronous evaluation.
err := l.workerPool.SubmitWithKey(n.NodeID(), func() {
  l.concurrentEvalFn(n, dependantCtx, tracer, parent)
})
```

I think it's slightly more idiomatic in Go for the caller to decide whether something is a blocking call or not.

Reply:
Yep, I agree it's more idiomatic and less indentation, so I've updated to it.